// nextjs-blog-lib/build/main/lib/content.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHtmlStringFromMarkdown = exports.createExcerpt = exports.rewriteMedias = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = require("path");
const react_1 = __importDefault(require("react"));
const server_1 = __importDefault(require("react-dom/server"));
/**
 * Rewrite all media images to public folder paths
 *
 * @ignore
 */
function rewriteMedias(baseFilePath, scope, markdown, mediaDirectory, relativeDirectory) {
    const mediaPaths = [];
    // Match images like `![alt](/path)` or `![alt](./path)`
    const regex = /!\[.*]\((\.\/|\/)(.*)\)/gm;
    let match;
    while ((match = regex.exec(markdown)) !== null) {
        const prefix = match[1];
        const mediaPath = match[2];
        mediaPaths.push({
            referencePath: mediaPath,
            oldAbsolutePath: path_1.join(baseFilePath, mediaPath),
            newPath: path_1.join(scope, mediaPath),
        });
        markdown = markdown.replace(`${prefix}${mediaPath}`, path_1.join(relativeDirectory, scope, mediaPath));
    }
    // Copy all media files to public folder
    for (const mediaPath of mediaPaths) {
        // Ensure the media scope folder exists
        const scopePath = path_1.join(mediaDirectory, scope);
        if (!fs_1.default.existsSync(scopePath)) {
            fs_1.default.mkdirSync(scopePath, { recursive: true });
        }
        const { oldAbsolutePath, newPath } = mediaPath;
        fs_1.default.copyFileSync(oldAbsolutePath, path_1.join(mediaDirectory, newPath));
    }
    return markdown;
}
exports.rewriteMedias = rewriteMedias;
/**
 * @ignore
 */
function createExcerpt(content, options) {
    var _a;
    const truncateIndex = content.indexOf(options.truncateMarker);
    if (truncateIndex !== -1) {
        return content.slice(0, truncateIndex);
    }
    // Find the first paragraph
    const firstParagraphIndex = (_a = content.match(/(.+)\n/gm)) === null || _a === void 0 ? void 0 : _a[0].length;
    if (firstParagraphIndex !== -1) {
        return content.slice(0, firstParagraphIndex);
    }
    return content.slice(0, options.truncateLength);
}
exports.createExcerpt = createExcerpt;
/**
 * Create a HTML string from a markdown string, and
 * support MDX syntax with imports. This is used to
 * generate server-side rendered HTML for blog posts.
 *
 * @ignore
 */
async function createHtmlStringFromMarkdown(markdown, options) {
    // Replace <!--.*--> with empty string
    markdown = markdown.replace(/<!--.*-->/g, "");
    const code = await options.mdx.compile(markdown);
    const Component = await toJsx(code, options);
    const rendered = String(server_1.default.renderToString(react_1.default.createElement(Component, { components: options.mdx.components })));
    return {
        rendered,
        code,
    };
}
exports.createHtmlStringFromMarkdown = createHtmlStringFromMarkdown;
/**
 * @ignore
 */
async function toJsx(code, options) {
    return options.mdx.run(code);
}
