// nextjs-blog-lib/build/main/lib/posts.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.groupPostsByTags = exports.getPreviousPost = exports.getNextPost = exports.getAllPostsByDate = exports.getAllPosts = exports.getPostSlugs = exports.getPostBySlug = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = require("path");
const gray_matter_1 = __importDefault(require("gray-matter"));
const content_1 = require("./content");
const util_1 = require("./util");
/**
 * Get a post by its slug.
 *
 * @param slug
 * @param options
 */
exports.getPostBySlug = util_1.memoize(async function (slug, options) {
    var _a, _b;
    const folderPath = path_1.join(options.postsDirectory, slug);
    const fullPath = path_1.join(folderPath, "index.mdx");
    const backupPath = path_1.join(folderPath, "index.md");
    let filePath;
    // Check if path is a folder or a file
    if (fs_1.default.lstatSync(folderPath).isFile()) {
        filePath = folderPath;
    }
    else if (fs_1.default.existsSync(fullPath)) {
        filePath = fullPath;
    }
    else {
        filePath = backupPath;
    }
    const fileContents = fs_1.default.readFileSync(filePath, "utf8");
    const newContents = options.rewriteMediaUrls
        ? content_1.rewriteMedias(folderPath, slug, fileContents, options.rewriteMediaUrls.mediaDirectory, options.rewriteMediaUrls.relativeDirectory)
        : fileContents;
    const { data, content } = gray_matter_1.default(newContents);
    const excerpt = content_1.createExcerpt(content, options);
    const { rendered: excerptRendered, code: excerptCode } = await content_1.createHtmlStringFromMarkdown(excerpt || "", options);
    const { rendered: contentRendered, code: contentCode } = await content_1.createHtmlStringFromMarkdown(content || "", options);
    const item = {
        slug: slug,
        date: slug.split("-").slice(0, 3).join("-"),
        title: (_a = data["title"]) !== null && _a !== void 0 ? _a : slug,
        frontmatter: data,
        contentRaw: content,
        contentHTML: contentRendered,
        contentCode,
        excerptRaw: excerpt,
        excerptHTML: excerptRendered,
        excerptCode,
        tags: (_b = data["tags"]) !== null && _b !== void 0 ? _b : [],
    };
    return item;
});
/**
 * Get a list of all post slugs.
 */
function getPostSlugs(options) {
    const fileNames = fs_1.default.readdirSync(options.postsDirectory);
    // Exclude files that start with a dot
    return fileNames.filter((fileName) => !fileName.startsWith("."));
}
exports.getPostSlugs = getPostSlugs;
/**
 * Get a list of all posts.
 */
exports.getAllPosts = util_1.memoize(async function (options) {
    const slugs = getPostSlugs(options);
    return await Promise.all(slugs.map((slug) => exports.getPostBySlug(slug, options)));
});
async function getAllPostsByDate(options) {
    const result = await exports.getAllPosts(options);
    return result.sort((post1, post2) => (post1.date > post2.date ? -1 : 1));
}
exports.getAllPostsByDate = getAllPostsByDate;
async function getNextPost(slug, options) {
    const allPosts = await getAllPostsByDate(options);
    const index = allPosts.map((post) => post.slug).indexOf(slug);
    if (index === allPosts.length - 1 || index === -1)
        return null;
    return allPosts[index + 1];
}
exports.getNextPost = getNextPost;
async function getPreviousPost(slug, options) {
    const allPosts = await getAllPostsByDate(options);
    const index = allPosts.map((post) => post.slug).indexOf(slug);
    if (index === 0 || index === -1)
        return null;
    return allPosts[index - 1];
}
exports.getPreviousPost = getPreviousPost;
exports.groupPostsByTags = util_1.memoize(async function (options) {
    const allPosts = await exports.getAllPosts(options);
    const result = {};
    allPosts.forEach((post) => {
        post.tags.forEach((tag) => {
            if (!result[tag])
                result[tag] = [];
            result[tag].push(post);
        });
    });
    // Sort each key-value
    Object.keys(result).forEach((key) => {
        result[key].sort((post1, post2) => (post1.date > post2.date ? -1 : 1));
    });
    return result;
});
